src/systems/systems/src/hydraulic/linear_actuator.rs

```diff
@@ -0,0 +1,1302 @@
+struct CoreHydraulicForce {
+    current_mode: LinearActuatorMode,
+    closed_valves_reference_position: Ratio,
+
+    active_hydraulic_damping_constant: f64,
+    fluid_compression_spring_constant: f64,
+    fluid_compression_damping_constant: f64,
+
+    max_flow: VolumeRate,
+    min_flow: VolumeRate,
+    flow_error_prev: VolumeRate,
+
+    bore_side_area: Area,
+    rod_side_area: Area,
+
+    last_control_force: Force,
+}
+impl CoreHydraulicForce {
+    const DEFAULT_I_GAIN: f64 = 0.2;
+    const DEFAULT_P_GAIN: f64 = 0.05;
+    const DEFAULT_FORCE_GAIN: f64 = 200000.;
+
+    #[allow(clippy::too_many_arguments)]
+    fn new(
+        init_position: Ratio,
+        active_hydraulic_damping_constant: f64,
+        fluid_compression_spring_constant: f64,
+        fluid_compression_damping_constant: f64,
+        max_flow: VolumeRate,
+        min_flow: VolumeRate,
+        bore_side_area: Area,
+        rod_side_area: Area,
+    ) -> Self {
+        Self {
+            current_mode: LinearActuatorMode::ClosedValves,
+            closed_valves_reference_position: init_position,
+            active_hydraulic_damping_constant,
+            fluid_compression_spring_constant,
+            fluid_compression_damping_constant,
+            max_flow,
+            min_flow,
+            flow_error_prev: VolumeRate::new::<gallon_per_second>(0.),
+            bore_side_area,
+            rod_side_area,
+            last_control_force: Force::new::<newton>(0.),
+        }
+    }
+    fn update_force(
+        &mut self,
+        required_position: Ratio,
+        requested_mode: LinearActuatorMode,
+        position_normalized: Ratio,
+        current_pressure: Pressure,
+        signed_flow: VolumeRate,
+        speed: Velocity,
+    ) -> Force {
+        self.update_actions(requested_mode, position_normalized);
+
+        self.force(
+            required_position,
+            position_normalized,
+            current_pressure,
+            signed_flow,
+            speed,
+        )
+    }
+
+    fn update_actions(&mut self, requested_mode: LinearActuatorMode, position_normalized: Ratio) {
+        match requested_mode {
+            LinearActuatorMode::ClosedValves => {
+                self.actions_from_current_to_closed_valves(position_normalized)
+            }
+            LinearActuatorMode::PositionControl => self.actions_from_current_to_position_control(),
+            LinearActuatorMode::ActiveDamping => self.actions_from_current_to_damping(),
+        }
+    }
+
+    fn actions_from_current_to_closed_valves(&mut self, position_normalized: Ratio) {
+        match self.current_mode {
+            LinearActuatorMode::ClosedValves => {}
+            LinearActuatorMode::PositionControl => {
+                self.go_to_close_control_valves(position_normalized);
+            }
+            LinearActuatorMode::ActiveDamping => {
+                self.go_to_close_control_valves(position_normalized);
+            }
+        }
```

Comment:
**suggestion**: Can be shortened to:
**note**: After this change, you could consider inlining `go_to_close_control_valves` and the other functions, as there is only one usage.

```suggestion
        match self.current_mode {
            LinearActuatorMode::ClosedValves => {}
            LinearActuatorMode::PositionControl | LinearActuatorMode::ActiveDamping => {
                self.go_to_close_control_valves(position_normalized);
            }
        }
```